src/ynEIGEN/AssetRegistry.sol

```diff
@@ -0,0 +1,358 @@
+    event AssetDeactivated(address indexed asset);
+    event PausedUpdated(bool paused);
+}
+
+/** @title AssetRegistry.sol
+ *  @dev This contract handles the strategy management for ynEigen asset allocations.
+ */
+ contract AssetRegistry is IAssetRegistry, AccessControlUpgradeable, ReentrancyGuardUpgradeable, IAssetRegistryEvents {
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  ERRORS  ------------------------------------------
+    //--------------------------------------------------------------------------------------
+
+    error UnsupportedAsset(IERC20 asset);
+    error Paused();
+    error AssetNotActiveOrNonexistent(address inactiveAsset);
+    error AssetBalanceNonZeroInPool(uint256 balanceInPool);
+    error AssetBalanceNonZeroInStrategyManager(uint256 balanceInStrategyManager);
+    error AssetNotFound(address absentAsset);
+    error ZeroAmount();
+    error ZeroAddress();
+    error LengthMismatch(uint256 length1, uint256 length2);
+    error AssetAlreadyActive(address asset);
+    error NoStrategyDefinedForAsset(IERC20 asset);
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  ROLES  -------------------------------------------
+    //--------------------------------------------------------------------------------------
+
+    bytes32 public constant ASSET_MANAGER_ROLE = keccak256("ASSET_MANAGER_ROLE");
+    bytes32 public constant PAUSER_ROLE = keccak256("PAUSER_ROLE");
+    bytes32 public constant UNPAUSER_ROLE = keccak256("UNPAUSER_ROLE");
+    
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  VARIABLES  ---------------------------------------
+    //--------------------------------------------------------------------------------------
+
+     /// @notice List of supported ERC20 asset contracts.
+    IERC20[] public assets;
+
+    mapping(IERC20 => AssetData) public _assetData;
+
+    bool public actionsPaused;
+
+    IRateProvider public rateProvider;
+    IEigenStrategyManager public eigenStrategyManager;
+    IynEigen ynEigen;
+
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  INITIALIZATION  ----------------------------------
+    //--------------------------------------------------------------------------------------
+
+    constructor() {
+       _disableInitializers();
+    }
+
+    struct Init {
+        IERC20[] assets;
+        IRateProvider rateProvider;
+        IEigenStrategyManager eigenStrategyManager;
+        IynEigen ynEigen;
+        address admin;
+        address pauser;
+        address unpauser;
+        address assetManagerRole;
+    }
+
+    function initialize(Init calldata init)
+        public
+        notZeroAddress(address(init.rateProvider))
```

Comment:
A modifier gets inlined when the code compiles. In this case, it will copy the modifier code into every function that uses it, increasing the deployment cost of the contract. For functions where the modifier is used multiple times, you might consider using a call to a function instead of a modifier to potentially reduce bytecode size and deployment costs.